gateway-sp-comms/src/communicator.rs

```diff
@@ -172,47 +180,130 @@ impl Communicator {
+    // to relay any information sent or received on that connection to the SP
+    // via UDP. SPs will continuously broadcast any serial console data, even if
+    // there is no attached client. Maybe this is fine, since the serial console
+    // shouldn't be noisy without a corresponding client driving it?
+    pub async fn serial_console_attach(
+        self: &Arc<Self>,
+        request: &mut http::Request<Body>,
         sp: SpIdentifier,
-        component: &SpComponent,
-    ) -> Result<Option<SerialConsoleContents>, Error> {
+        component: SpComponent,
+        sp_ack_timeout: Duration,
+    ) -> Result<http::Response<Body>, Error> {
         let port = self.id_to_port(sp)?;
-        Ok(self.recv_handler.serial_console_contents(port, component))
+
+        if !request
+            .headers()
+            .get(header::CONNECTION)
+            .and_then(|hv| hv.to_str().ok())
+            .map(|hv| {
+                hv.split(|c| c == ',' || c == ' ')
+                    .any(|vs| vs.eq_ignore_ascii_case("upgrade"))
+            })
+            .unwrap_or(false)
+        {
+            return Err(Error::BadWebsocketConnection(
+                "expected connection upgrade",
+            ));
+        }
+        if !request
+            .headers()
+            .get(header::UPGRADE)
+            .and_then(|v| v.to_str().ok())
+            .map(|v| {
+                v.split(|c| c == ',' || c == ' ')
+                    .any(|v| v.eq_ignore_ascii_case("websocket"))
+            })
+            .unwrap_or(false)
+        {
+            return Err(Error::BadWebsocketConnection(
+                "unexpected protocol for upgrade",
+            ));
+        }
+        if request
+            .headers()
+            .get(header::SEC_WEBSOCKET_VERSION)
+            .map(|v| v.as_bytes())
+            != Some(b"13")
+        {
+            return Err(Error::BadWebsocketConnection(
+                "missing or invalid websocket version",
+            ));
+        }
+        let accept_key = request
+            .headers()
+            .get(header::SEC_WEBSOCKET_KEY)
+            .map(|hv| hv.as_bytes())
+            .map(|key| handshake::derive_accept_key(key))
+            .ok_or(Error::BadWebsocketConnection("missing websocket key"))?;
+
+        self.recv_handler.serial_console_attach(
+            Arc::clone(self),
+            port,
+            component,
+            sp_ack_timeout,
+            upgrade::on(request),
+        )?;
+
+        // `.body()` only fails if our headers are bad, which they aren't
+        // (unless `hyper::handshake` gives us a bogus accept key?), so we're
+        // safe to unwrap this
+        Ok(http::Response::builder()
+            .status(http::StatusCode::SWITCHING_PROTOCOLS)
+            .header(header::CONNECTION, "Upgrade")
+            .header(header::UPGRADE, "websocket")
+            .header(header::SEC_WEBSOCKET_ACCEPT, accept_key)
+            .body(Body::empty())
+            .unwrap())
```

Comment:
I know you got this from the propolis console (@luqmana, very cool! I should have looked sooner) so do you two think this is something that dropshot should help with? While it's not OpenAPI, Async API seems to be compatible so we could emit a schema of expected websocket messages in the "OpenAPI" json dropshot outputs. We could also update progenitor to maybe make for a nice generated interface... though I'm not sure what that would look like exactly.

Reply:
I'd vote yes, somewhat naively. At a minimum it doesn't seem ideal that we're picking out HTTP headers by hand to confirm the request looks like a websocket one. I'm not sure what the API would look like - would we be describing the messages sent in the websocket, or just the fact that this endpoint exposes a websocket?

Reply:
Yea, kinda waffled back and forth on adding it to dropshot proper. But now that we have more and more uses of the same pattern I think we should seriously consider what that sort of integration might look like. (@lifning too will need something along these lines in nexus afaik)

Reply:
Thanks @luqmana and @jgallagher; I'll poke at this a bit now that I'm back